From 96f49c947c23c10002e15d1ef6f57bed26ea1399 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:05:21 -0400
Subject: [usb] oti6858: only wake up MSR queue on changes

Message-id: <1410887463-20674-250-git-send-email-dzickus@redhat.com>
Patchwork-id: 95316
O-Subject: [RHEL7 PATCH 249/591] USB: oti6858: only wake up MSR queue on changes
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 1acc36e9c8dcd44f0978c9270e24daa83f343f85
Author: Johan Hovold <jhovold@gmail.com>
Date:   Thu Jan 2 22:49:34 2014 +0100

    USB: oti6858: only wake up MSR queue on changes

    Only wake up MSR wait queue on actual modem-status changes.

    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/serial/oti6858.c b/drivers/usb/serial/oti6858.c
index 1dea599..9ea15e0 100644
--- a/drivers/usb/serial/oti6858.c
+++ b/drivers/usb/serial/oti6858.c
@@ -103,6 +103,7 @@ struct oti6858_control_pkt {
 #define TX_BUFFER_EMPTIED 0x09
  u8 pin_state;
 #define PIN_MASK  0x3f
+#define PIN_MSR_MASK  0x1b
 #define PIN_RTS   0x20 /* output pin */
 #define PIN_CTS   0x10 /* input pin, active low */
 #define PIN_DSR   0x08 /* input pin, active low */
@@ -739,8 +740,11 @@ static void oti6858_read_int_callback(struct urb *urb)
   }
 
   if (!priv->transient) {
-   if (xs->pin_state != priv->status.pin_state)
+   u8 delta = xs->pin_state ^ priv->status.pin_state;
+
+   if (delta & PIN_MSR_MASK)
     wake_up_interruptible(&port->port.delta_msr_wait);
+
    memcpy(&priv->status, xs, OTI6858_CTRL_PKT_SIZE);
   }
 
-- 
1.7.1